plugins/com.google.cloud.tools.eclipse.appengine.libraries.ui/src/com/google/cloud/tools/eclipse/appengine/libraries/ui/CloudLibrariesPage.java

```diff
@@ -18,106 +18,197 @@
+        BuildPath.saveLibraryList(project, libraries, new NullProgressMonitor());
+        if (newEntry == null) {
+          // container-editing only refreshes the content if new
+          BuildPath.runContainerResolverJob(project);
+        }
+      }
+      return true;
+    } catch (CoreException ex) {
+      StatusUtil.setErrorStatus(this, "Error updating container definition", ex); //$NON-NLS-1$
+      return false;
+    }
   }
 
-  @Override
-  public IClasspathEntry getSelection() {
-    // Since this class implements IClasspathContainerPageExtension2,
-    // Eclipse calls getNewContainers instead.
-    logger.log(Level.WARNING, "Unexpected call to getSelection()");
-    return null;
+  @VisibleForTesting
+  List<Library> getVisibleLibraries() {
+    List<Library> visible = new ArrayList<>();
+    for (LibrarySelectorGroup librariesSelector : librariesSelectors) {
+      visible.addAll(librariesSelector.getAvailableLibraries());
+    }
+    return visible;
   }
 
-  @Override
-  public void setSelection(IClasspathEntry containerEntry) {
-    // todo can we use the containerEntry to tick the checkboxes in the library selector group?
+  /**
+   * Return the list of selected libraries.
+   */
+  @VisibleForTesting
+  List<Library> getSelectedLibraries() {
+    List<Library> selectedLibraries = new ArrayList<>();
+    for (LibrarySelectorGroup librariesSelector : librariesSelectors) {
+      selectedLibraries.addAll(librariesSelector.getSelectedLibraries());
+    }
+    return selectedLibraries;
   }
 
-  @Override
-  public void initialize(IJavaProject project, IClasspathEntry[] currentEntries) {
-    this.project = project;
+  @VisibleForTesting
+  void setSelectedLibraries(List<Library> selectedLibraries) {
+    initialSelection = new ArrayList<>(selectedLibraries);
+    List<Library> remaining = new ArrayList<>(selectedLibraries);
+    if (!librariesSelectors.isEmpty()) {
+      for (LibrarySelectorGroup librarySelector : librariesSelectors) {
+        librarySelector.setSelection(new StructuredSelection(initialSelection));
+        remaining.removeAll(librarySelector.getSelectedLibraries());
+      }
+      if (!remaining.isEmpty()) {
+        logger.log(Level.WARNING,
+            "Discarding libraries that aren't availble from library container definition: " //$NON-NLS-1$
+                + remaining);
+      }
+    }
   }
 
   @Override
-  public IClasspathEntry[] getNewContainers() {
-    List<Library> libraries = new ArrayList<>(librariesSelector.getSelectedLibraries());
-    if (libraries == null || libraries.isEmpty()) {
-      return null;
+  public void setSelection(IClasspathEntry containerEntry) {
+    // Although null means "new entry", we only support a single entry
+    // and since we use a fixed container path, we treat null as edit-current-entry
+    // FIXME: might be useful to read in current dependencies for maven projects
+    oldEntry = containerEntry;
+    if (project == null || isMavenProject) {
```

Comment:
`project` can never be null at this point. `project` --> `oldEntry`? Even so, there is no need for a class field.

Reply:
Hmm, there was a reason, but I don't recollect and the tests aren't failing.  The comment is also out of date, so I'll toss that too.